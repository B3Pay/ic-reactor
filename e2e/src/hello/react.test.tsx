import createReActor from "@ic-reactor/react"
import React from "react"
import renderer, { act } from "react-test-renderer"
import { canisterId, createActor } from "../declarations/hello_actor"

describe("React Test", () => {
  it("should initialize", async () => {
    const { ReActorProvider, initialize, useReActor } = createReActor(
      (agent) =>
        createActor(canisterId, {
          agent,
        }),
      {
        initializeOnMount: false,
      }
    )

    const TestInitialize = () => {
      const { initialized, initializing } = useReActor()
      return (
        <div>
          <span>
            {initializing
              ? "Initializing"
              : initialized
              ? "Initialized"
              : "Not initialized"}
          </span>
          <button onClick={() => initialize()}>Initialize</button>
        </div>
      )
    }

    let screen = renderer.create(
      <ReActorProvider>
        <TestInitialize />
      </ReActorProvider>
    )

    const initializeStatus = () =>
      screen.root.findAllByType("span")[0].props.children
    const initializeButton = () => screen.root.findAllByType("button")[0]

    expect(screen.toJSON()).toMatchSnapshot()

    expect(initializeStatus()).toEqual("Not initialized")

    await act(() => initializeButton().props.onClick())

    expect(initializeStatus()).toEqual("Initialized")

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it("should call query method", async () => {
    const { ReActorProvider, useQueryCall } = createReActor((agent) =>
      createActor(canisterId, {
        agent,
      })
    )

    const TestComponent = () => {
      const { recall, data, loading } = useQueryCall({
        functionName: "greet",
        args: ["Query Call"],
      })

      return (
        <div>
          <button onClick={() => recall()}>Say Hello</button>
          <span>{loading ? "Loading..." : data ? data : "Ready To call"}</span>
        </div>
      )
    }

    let screen = renderer.create(
      <ReActorProvider>
        <TestComponent />
      </ReActorProvider>
    )

    const data = () => screen.root.findAllByType("span")[0]

    const button = () => screen.root.findAllByType("button")[0]

    expect(screen.toJSON()).toMatchSnapshot()

    expect(data().props.children).toEqual("Ready To call")

    await act(() => button().props.onClick())

    expect(data().props.children).toEqual("Hello, Query Call!")

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it("should call update method", async () => {
    const { ReActorProvider, useUpdateCall } = createReActor((agent) =>
      createActor(canisterId, {
        agent,
      })
    )
    const TestComponent = () => {
      const { call, data, loading } = useUpdateCall({
        functionName: "greet_update",
        args: ["Update Call"],
      })

      return (
        <div>
          <button onClick={() => call()}>Say Hello</button>
          <span>{loading ? "Loading..." : data ? data : "Ready To call"}</span>
        </div>
      )
    }

    let screen = renderer.create(
      <ReActorProvider>
        <TestComponent />
      </ReActorProvider>
    )

    const data = () => screen.root.findAllByType("span")[0]

    const button = () => screen.root.findAllByType("button")[0]

    expect(screen.toJSON()).toMatchSnapshot()

    expect(data().props.children).toEqual("Ready To call")

    await act(() => button().props.onClick())

    expect(data().props.children).toEqual("Hello, Update Call!")

    expect(screen.toJSON()).toMatchSnapshot()
  })
})
