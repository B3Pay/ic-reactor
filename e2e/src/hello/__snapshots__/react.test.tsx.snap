// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Test should call query method 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`React Test should call query method 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Hello, Query Call!
  </span>
</div>
`;

exports[`React Test should call update method 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`React Test should call update method 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Hello, Update Call!
  </span>
</div>
`;

exports[`React Test should initialize 1`] = `
<div>
  <span>
    Not initialized
  </span>
  <button
    onClick={[Function]}
  >
    Initialize
  </button>
</div>
`;

exports[`React Test should initialize 2`] = `
<div>
  <span>
    Initialized
  </span>
  <button
    onClick={[Function]}
  >
    Initialize
  </button>
</div>
`;

exports[`React Test should query 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`React Test should query 2`] = `
<div>
  <button
    onClick={[Function]}
  >
    Say Hello
  </button>
  <span>
    Ready To call
  </span>
</div>
`;
